### PR TITLE
Adding cmake option for disabling preconditions, docs and CI pipeline

### PR DESCRIPTION
**Pull request by vhvb1989**
Integrating NO_PRECONDITION generate, build and unit test running in CI

### CODE REVIEW

**Comment by vhvb1989** [comment]
https://github.com/Azure/azure-sdk-for-c/issues/553

File: sdk/iot/pnp/src/az_iot_pnp_client_telemetry.c [hunk 1]
```diff
@@ -58,6 +58,7 @@ AZ_NODISCARD az_result az_iot_pnp_client_telemetry_get_publish_topic(
   AZ_PRECONDITION_VALID_SPAN(mqtt_topic, 0, false);
   AZ_PRECONDITION_IS_NULL(reserved);
   AZ_PRECONDITION_NOT_NULL(out_mqtt_topic);
+  (void)reserved;
```
Comment (vhvb1989) [inline sdk/iot/pnp/src/az_iot_pnp_client_telemetry.c:61]: @CIPop  @danewalton 
Do we need to have all this methods with those unused parameters?  Why can't we just remove those parameters?
I have patched this for PRECONDITION=OFF, but I wonder if you can just have those removed if you don't use them

Comment (danewalton) [reply sdk/iot/pnp/src/az_iot_pnp_client_telemetry.c:61]: I believe it is a placeholder for future fields. @jspaith added it for pnp necessities.

File: CMakeLists.txt [hunk 2]
```diff
@@ -10,6 +10,12 @@ option(BUILD_CURL_TRANSPORT "Build internal http transport implementation with C
 option(UNIT_TESTING "Build unit test projects" OFF)
 option(UNIT_TESTING_MOCK_ENABLED "wrap PAL functions with mock implementation for tests" OFF)
 option(BUILD_PAHO_TRANSPORT "Build IoT Samples with Paho MQTT support" OFF)
+option(BUILD_PRECONDITIONS "Build SDK with preconditions enabled" ON)
+
+#enable mock functions with link option -ld
```
Comment (ahsonkhan) [inline CMakeLists.txt:15]: comment mismatch

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 3]
```diff
@@ -7,7 +7,14 @@ jobs:
...
     matrix:
-      # Build with no dependencies at all (No samples)
+      # Build with no dependencies at all (No samples) - PRECONDITIONS ON
+      # WARNINGS_AS_ERRORS          ON
+      # BUILD_CURL_TRANSPORT        OFF
+      # UNIT_TESTING                OFF
+      # UNIT_TESTING_MOCK_ENABLED   OFF
+      # BUILD_PAHO_TRANSPORT        OFF
+      # BUILD_PRECONDITIONS         ON
+      # AZ_PLATFORM_IMPL            AZ_PLATFORM_IMPL_NONE
```
Comment (ahsonkhan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:17]: Nice!

File: eng/pipelines/templates/jobs/archetype-sdk-client.yml [hunk 4]
```diff
@@ -55,7 +69,14 @@ jobs:
         VCPKG_DEFAULT_TRIPLET: 'x64-osx'
         build.args: ' -DBUILD_CURL_TRANSPORT=ON -DAZ_PLATFORM_IMPL=POSIX'
 
-      # Build with sample dependencies and unit testing [curl for transport and cmoka]
+      # Build with sample dependencies and unit testing [curl for transport and cmoka]  - PRECONDITIONS ON
```
Comment (ahsonkhan) [inline eng/pipelines/templates/jobs/archetype-sdk-client.yml:72]: nit: typo - cmocka - find/replace in a bunch of places.
```suggestion
      # Build with sample dependencies and unit testing [curl for transport and cmocka]  - PRECONDITIONS ON
```

File: sdk/core/core/README.md [hunk 5]
```diff
@@ -148,7 +148,7 @@ The public SDK functions validate the arguments passed to them to ensure that th
 
 To override the default behavior, implement a function matching the `az_precondition_failed_fn` function signature and then, in your application's initialization (before calling any Azure SDK function), call `az_precondition_failed_set_callback` passing it the address of your function. Now, when any Azure SDK function detects a precondition failure, it will invoke your callback instead. You might override the callback to attach a debugger or perhaps to reboot the device rather than allowing it to continue running with unpredictable behavior.
 
-Also, if you define the `NO_PRECONDITION_CHECKING` symbol when compiling the SDK code, all of the Azure SDK precondition checking will be excluded, making the binary code smaller and faster. We recommend doing this before you ship your code.
+Also, if you define the `NO_PRECONDITION_CHECKING` symbol when compiling the SDK code (or adding option -DBUILD_PRECONDITIONS=OFF with cmake), all of the Azure SDK precondition checking will be excluded, making the binary code smaller and faster. We recommend doing this before you ship your code.
```
Comment (ahsonkhan) [inline sdk/core/core/README.md:151]: nit: Rename the option to be `BUILD_WITH_PRECONDITIONS`

**Comment by ahsonkhan** [comment]
Btw, @vhvb1989 - if you use keywords like fix/fxes with the issue number as part of your PR description, the issue get's auto-closed when the PR is merged (with a link to the PR that closed the issue showing up in the issue itself). That way you don't need to manually close the issue.

https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

**Pull request merged**